Some 2.2 performance improvements.

rryan opened the pull request:
This is a grab bag of 2.2 performance tweaks, aimed at further improving performance on macOS after  #1809. These were mostly discovered with a profiler while trying to get Mixxx to use less than 80% of a 4-core 2017 macbook pro when sitting idle ;).

rryan committed:
Cache creation of QIcons in LibraryFeatures.

This showed up in profiling as a hotspot after disabling waveforms.

rryan committed:
Render and swap WSpinny widgets at the same time as waveform widgets.

WSpinny widgets are QGLWidgets with autoBufferSwap enabled. This means that
after they are painted, they automatically run swapBuffers(). This can block the
main thread until a vsync occurs, and generally interferes with the
VSyncThread's efforts to render all the waveform widgets in one step then swap
them all at once.

rryan committed:
Fix EngineBuffer indicator update rate variables.

The constants suggest the update rate was 10 Hz, but due to a sample
vs. frames difference it is actually 20 Hz. This fixes the constant
values without changing behavior.

rryan committed:
Don't render hidden or unexposed QGLWidgets.

Fixes LP Bug #1801636.

rryan committed:
Add override annotations in GLSLWaveformWidget.

rryan committed:
Move GuiTick processing from VSyncThread to WaveformWidgetFactory.

GuiTick is delivering more and more render signals to widgets in the main
thread, and when emitted from the VSyncThread, we're putting a heavy load on the
Qt event queue. When these signals are emitted from
WaveformWidgetRenderer::render(), they are emitted from the main thread, and so
can be delivered directly rather than queued.

Be-ing commented:
Thanks for working on this. It may be helpful to set a specific time in the release cycle before the release candidate stage to focus on performance improvements driven by profiling.

rryan commented:
> Thanks for working on this. It may be helpful to set a specific time in the release cycle before the release candidate stage to focus on performance improvements driven by profiling.

I agree.. I'm only working on this now because while Mixxx 2.2 is usable with 2 decks, when you go up to 4 decks with 4 spinnies, it's unable to perform smoothly on ridiculously high-spec hardware (4-core 2017 macbook pro)

Be-ing commented on src/waveform/guitick.cpp:
```diff
@@ -10,7 +10,7 @@ GuiTick::GuiTick(QObject* pParent)
 }
 
-// this is called from the VSyncThread
+// this is called from WaveformWidgetFactory::render
```
It would help to add a bit more to the comment: "This is called from WaveformWidgetFactory::render in the main thread"

Be-ing reviewed (changes requested):
CPU usage for me goes from ~60% with one deck playing on the 2.2 branch to ~45% with this PR. This is with a quad core Intel Core i7-8550U CPU.

Be-ing commented:
I'm wary of merging the changes in 84054a9 regarding communication between threads at this point. Maybe we could merge the rest of this PR to 2.2 and cherry pick 84054a9 to master?

rryan commented:
We can update the desired rate to 2x while keeping the fix, but note that
that increses CPU usage for me a fair bit on macOS (rendering text at 20
fps isn't cheap, and causes the rest of the GUI to be repainted to some
extent). Maybe increasing the rate by 1.5x instead of 2 would be ok?

On Tue, Nov 13, 2018, 11:26 PM Daniel Schürmann <notifications@github.com>
wrote:

> *@daschuer* commented on this pull request.
> ------------------------------
>
> In src/engine/enginebuffer.cpp
> <https://github.com/mixxxdj/mixxx/pull/1897#discussion_r233336093>:
>
> > @@ -1235,7 +1235,7 @@ void EngineBuffer::updateIndicators(double speed, int iBufferSize) {
>
>      // Update indicators that are only updated after every
>      // sampleRate/kiUpdateRate samples processed.  (e.g. playposSlider)
> -    if (m_iSamplesCalculated > (m_pSampleRate->get() / kiPlaypositionUpdateRate)) {
> +    if (m_iSamplesCalculated > (kSamplesPerFrame * m_pSampleRate->get() / kiPlaypositionUpdateRate)) {
>
> This fixes a documentation error, but effectively slows down the update
> rate significant. Is this necessary to fix this bug?
> Before we had a 50 ms rate, now we have 100 ms and 400 ms for BPM. The old
> values have just worked. Does it really help? Else I prefere only fix the
> code documentation.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mixxxdj/mixxx/pull/1897#discussion_r233336093>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AABnnzzBmO-xkH9x9lKo7Stzob81gPd2ks5uu8WwgaJpZM4YagWS>
> .
>

daschuer commented:
Any update rate works for me that solves you issue and does not look odd.
My worries where to make a change here, just because there was an issue in the code comments.

daschuer commented:
LGTM
Just tested it with no notable regressions.

Be-ing commented:
Hmm, I tested the commit before 84054a9 and CPU usage with one deck playing was the same as without this branch (~60%). I'm amazed what a big difference that change makes. I wonder if there are other relatively low hanging fruit for optimization like that where we are emitting signals between threads with a high frequency...

rryan commented:
I feel they're all pretty safe except for 32056875ff8faaa128cdb3550b1557e7de489224, given the issue with Windows.

Be-ing commented:
I agree. Can you rebase to remove that commit from this branch?

rryan commented on src/waveform/waveformwidgetfactory.cpp:
```diff
@@ -535,6 +535,10 @@ void WaveformWidgetFactory::render() {
 
+        // WSpinnys are also double-buffered QGLWidgets, like all the waveform
+        // renderers. Render all the WSpinny widgets now.
+        emit(renderSpinnies());
```
it might be confusing, because some waveform widgets are also QGLWidgets and this signal doesn't affect them

rryan committed:
Change EngineBuffer indicator update rate to 15 Hz.

At 20 Hz, the GUI spends a lot of time re-drawing the WNumberPos
widgets (which can trigger re-renders of the whole UI, since their
parents need to be re-drawn).

rryan commented on src/library/recording/recordingfeature.cpp:
```diff
@@ -32,7 +33,7 @@ QVariant RecordingFeature::title() {
 }
 
 QIcon RecordingFeature::getIcon() {
```
QIcon is PIMPL class like most other Qt classes, so it's ~cheap to return copies.

rryan commented on src/engine/enginebuffer.cpp:
```diff
@@ -1235,7 +1235,7 @@ void EngineBuffer::updateIndicators(double speed, int iBufferSize) {
     // Update indicators that are only updated after every
     // sampleRate/kiUpdateRate samples processed.  (e.g. playposSlider)
-    if (m_iSamplesCalculated > (m_pSampleRate->get() / kiPlaypositionUpdateRate)) {
+    if (m_iSamplesCalculated > (kSamplesPerFrame * m_pSampleRate->get() / kiPlaypositionUpdateRate)) {
```
I fixed this commit to only fix the documentation, and added a separate commit to change the update rate from 20 Hz to 15 Hz (instead of 10 Hz). Does that work?

rryan commented on src/waveform/waveformwidgetfactory.cpp:
```diff
@@ -517,8 +542,14 @@ void WaveformWidgetFactory::render() {
             for (int i = 0; i < m_waveformWidgetHolders.size(); i++) {
+                WaveformWidgetAbstract* pWaveformWidget = m_waveformWidgetHolders[i].m_waveformWidget;
+                // Don't bother doing the pre-render work if we aren't going to
+                // render this widget. There is a tiny race condition here.
```
shouldRenderWaveform could become true in the time between this for loop executing and the for loop below which calls render, at which point we'd call render on a waveform that hasn't has prerender called on it.  I updated the comment.

rryan committed:
Update GuiTick comment.
